### PR TITLE
limine: init at 4.20230120.0

### PR DESCRIPTION
**Pull request by 48cf**
###### Description of changes

This patch introduces support for the [Limine bootloader](https://github.com/limine-bootloader/limine/). Limine is a modern, advanced, portable, multiprotocol bootloader. It supports the following architectures (currently only the first two are supported by this PR due to lack of testing):

- IA-32 (32-bit x86)
- x86_64
- aarch64 (arm64)

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
  - [x] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by N00byEdge** [comment]
This would be really cool and seems to work well on my computer!

File: pkgs/tools/bootloaders/limine/default.nix
Comment (dasJ) [inline pkgs/tools/bootloaders/limine/default.nix:3]: The use of `pkgs` is discouraged, you get all packages as parameters anyway.

File: pkgs/tools/bootloaders/limine/default.nix
Comment (dasJ) [inline pkgs/tools/bootloaders/limine/default.nix:4]:
```suggestion
  pname = "limine";
```

File: pkgs/tools/bootloaders/limine/default.nix
Comment (dasJ) [inline pkgs/tools/bootloaders/limine/default.nix:6]: Please use `fetchFromGitHub` here

File: pkgs/tools/bootloaders/limine/default.nix
Comment (dasJ) [inline pkgs/tools/bootloaders/limine/default.nix:11]: Again, no `pkgs` please. Also the auto* packages should already be here along with the binutils

File: pkgs/tools/bootloaders/limine/default.nix
Comment (dasJ) [inline pkgs/tools/bootloaders/limine/default.nix:15]: You can use `clangStdenv` instead of having a package that depends on both gcc (from stdenv) and clang

File: pkgs/tools/bootloaders/limine/default.nix
Comment (dasJ) [inline pkgs/tools/bootloaders/limine/default.nix:19]: Please set `configureFlags` instead

File: pkgs/tools/bootloaders/limine/default.nix
Comment (dasJ) [inline pkgs/tools/bootloaders/limine/default.nix:22]: That entire phase should be unnecessary when using the default configure phase

**Review by dasJ — CHANGES_REQUESTED** [review]

File: pkgs/tools/bootloaders/limine/default.nix
Comment (czapek1337) [reply pkgs/tools/bootloaders/limine/default.nix:6]: I used `fetchurl` here because I need the release tarball instead of the repository itself. The tarballs have all the dependencies needed for building the package.

File: pkgs/tools/bootloaders/limine/default.nix
Comment (czapek1337) [reply pkgs/tools/bootloaders/limine/default.nix]: Done.

File: pkgs/tools/bootloaders/limine/default.nix
Comment (czapek1337) [reply pkgs/tools/bootloaders/limine/default.nix]: Done.

File: pkgs/tools/bootloaders/limine/default.nix
Comment (czapek1337) [reply pkgs/tools/bootloaders/limine/default.nix]: Done.

File: pkgs/tools/bootloaders/limine/default.nix
Comment (czapek1337) [reply pkgs/tools/bootloaders/limine/default.nix]: Done.

File: pkgs/tools/bootloaders/limine/default.nix
Comment (czapek1337) [reply pkgs/tools/bootloaders/limine/default.nix]: Done.

File: pkgs/tools/bootloaders/limine/default.nix
Comment (czapek1337) [reply pkgs/tools/bootloaders/limine/default.nix]: Done.

**Review by dasJ — COMMENTED** [review]
Happy with the package now, but I cannot judge the python code. Maybe @mweinelt can? (if you have the capacity)

File: nixos/modules/system/boot/loader/limine/limine-install.py
Comment (mweinelt) [inline nixos/modules/system/boot/loader/limine/limine-install.py:1]: I'm going to be taking a look, but it will take a while to go through this.

Would you be opposed to adding typing hints?

File: nixos/modules/system/boot/loader/limine/limine-install.py
Comment (mweinelt) [inline nixos/modules/system/boot/loader/limine/limine-install.py:22]: This class seems CC-BY-SA licensed and the script as a whole could be a deritvative, but this repo requires MIT licensing.

https://stackoverflow.com/a/13520518

Comment (czapek1337) [reply nixos/modules/system/boot/loader/limine/limine-install.py:22]: Oh I missed that, that code was not meant to make it into this version, whoops. What should I do about it? Is a comment pointing to that answer enough? I'm really not that good when it comes to licensing stuff.

Comment (mweinelt) [reply nixos/modules/system/boot/loader/limine/limine-install.py:22]: CC-BY-SA requires to redistribution under the same license. The licensing of this repository is MIT, see https://github.com/NixOS/nixpkgs/blob/master/COPYING. I would say that is incompatible.

**Comment by mweinelt** [comment]
For a component as important as a bootloader I would like to see a NixOS test, compare the ones for grub and system-boot.

File: nixos/modules/system/boot/loader/limine/limine-install.py
Comment (czapek1337) [reply nixos/modules/system/boot/loader/limine/limine-install.py:1]: No, I'll get to it as soon as I can.

File: nixos/modules/system/boot/loader/limine/limine-install.py
Comment (alyssais) [reply nixos/modules/system/boot/loader/limine/limine-install.py:22]: You can't take a CC BY-SA work and distribute it in the MIT license, but we do bundle lots of non-MIT code in Nixpkgs, and I don't think there's anything stopping anybody mixing the two, as long as they can follow the terms of both licenses.   All the patches, for example (we have a note about this in the README), and occasionally we'll just include replacement versions of files not as patches (which technically the README note doesn't cover).

So it's complicated.  What we really need in Nixpkgs is per-file license tracking, rather than the misleading idea that "nixpkgs is MIT" or whatever, which isn't really true.  Are we okay with a CC BY-SA program included in Nixpkgs?  I don't see why we wouldn't be, since we already include so many GPL etc files, which have similar conditions.  But we should be clear about it, and probably revise the note in the README to say that in addition to the thing about patches, there are other exceptions which are clearly noted.

Comment (alyssais) [reply nixos/modules/system/boot/loader/limine/limine-install.py:22]: (But if you don't want to open that can of worms it might be easier to just not include this code. :))

File: maintainers/maintainer-list.nix
Comment (czapek1337) [reply maintainers/maintainer-list.nix]: Done.

**Comment by czapek1337** [comment]
> For a component as important as a bootloader I would like to see a NixOS test, since I won't test this on any production machine I have. See the ones for grub and system-boot.

Done. I'm not sure what tests there should be so for now I only added a "Machine boots correctly" which passes locally.

**Comment by czapek1337** [comment]
> Happy with the package now, but I cannot judge the python code. [...]

Could you please approve the changes then? You still have "changes requested" 😅

**Comment by czapek1337** [comment]
@mweinelt any chance you could take a look at the Python script? Also @dasJ your "changes requested" review is still there, since you said you were happy with the rest of the package could you please approve the changes? 🙂

**Comment by czapek1337** [comment]
I updated the package to the latest version and moved the changelog entry from 22.11 to 23.05. All previous concerns were addressed and I think it's ready for merge 🎉

**Comment by Juliaria08** [comment]
@czapek1337 What happened? Why wasn't it reviewed? (if you don't mind telling), as I haven't seen any other limine package requests nor the presence of limine in nixos' repo.

**Comment by czapek1337** [comment]
> @czapek1337 What happened? Why wasn't it reviewed? (if you don't mind telling), as I haven't seen any other limine package requests nor the presence of limine in nixos' repo.

No idea, I know as much as you do. I tried bumping it a few times but nada, so I closed the PR since it wasn’t going anywhere. Unfortunately it is blocked on code owners (@mweinelt and @dasJ) so unless they approve we can’t proceed. I’d be happy to reopen the PR though!

**Comment by Juliaria08** [comment]
> > @czapek1337 What happened? Why wasn't it reviewed? (if you don't mind telling), as I haven't seen any other limine package requests nor the presence of limine in nixos' repo.
> 
> No idea, I know as much as you do. I tried bumping it a few times but nada, so I closed the PR since it wasn’t going anywhere. Unfortunately it is blocked on code owners (@mweinelt and @dasJ) so unless they approve we can’t proceed. I’d be happy to reopen the PR though!

Oh, thanks for the information. @dasJ Said he approved the nix configs but they haven't marked it as approved in github so it would still be blocking... (as what practically counts for the merging (in GitHub's UI) is that two users which have access or are in a group have approved the PR right?)

**Comment by programmerlexi** [comment]
@48cf may i use this to make a new limine nixos module?

**Pull request closed without merging**
